File: sdk/typescript/src/files.ts
import { access, lstat, mkdir, readdir, readFile } from "node:fs/promises";
import path from "node:path";
import { create, extract } from "tar";
import { createRequire } from "node:module";
interface DockerIgnore { add(patterns: string): DockerIgnore; ignores(path: string): boolean }
const dockerignore = createRequire(import.meta.url)("@balena/dockerignore") as (options: { ignorecase: boolean }) => DockerIgnore;

export const MAX_TRANSFER = 512 * 1024 * 1024;
export async function readBounded(response: Response, limit = MAX_TRANSFER): Promise<Uint8Array> {
  if (!response.body) return new Uint8Array();
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = []; let size = 0;
  try {
    while (true) {
      const { value, done } = await reader.read();
      if (done) break;
      size += value.byteLength;
      if (size > limit) throw new Error("Download exceeds transfer limit");
      chunks.push(value);
    }
    return Buffer.concat(chunks);
  } finally {
    await reader.cancel();
    reader.releaseLock();
  }
}

export async function exists(name: string): Promise<boolean> {
  try { await access(name); return true; } catch { return false; }
}
export async function pack(local: string, context = false): Promise<{ data: Uint8Array; dockerfile?: string }> {
  const source = path.resolve(local);
  const stat = await lstat(source);
  if (stat.isSymbolicLink()) throw new Error("Transfers do not follow symlinks");
  const root = stat.isDirectory() ? source : path.dirname(source);
  const dockerfile = context ? stat.isDirectory() ? "Dockerfile" : path.basename(source) : undefined;
  if (dockerfile && !await exists(path.join(root, dockerfile))) throw new Error("Project context must contain a Dockerfile");
  const ignore = dockerignore({ ignorecase: false });
  if (context) {
    let ignorePath = path.join(root, `${dockerfile}.dockerignore`);
    if (!await exists(ignorePath)) ignorePath = path.join(root, ".dockerignore");
    if (await exists(ignorePath)) ignore.add(await readFile(ignorePath, "utf8"));
  }
  const entries: string[] = [];
  let total = 0;
  async function walk(directory: string): Promise<void> {
    for (const name of (await readdir(directory)).sort()) {
      const absolute = path.join(directory, name);
      const relative = path.relative(root, absolute).split(path.sep).join("/");
      if (context && relative.split("/").some(p => [".git", ".venv", ".cache", "node_modules"].includes(p))) continue;
      const info = await lstat(absolute);
      const excluded = context && relative !== dockerfile && ignore.ignores(relative);
      if (info.isDirectory()) {
        // Walk excluded directories too: Docker negations can re-include descendants.
        await walk(absolute);
        if (!excluded && (await readdir(absolute)).length === 0) entries.push(relative);
      } else if (!excluded) {
        if (!info.isFile()) throw new Error(`Cannot transfer symlink or special file: ${relative}`);
        total += info.size;
        if (total > MAX_TRANSFER) throw new Error("Transfer exceeds 512 MiB");
        entries.push(relative);
      }
    }
  }
  if (stat.isDirectory() || context) await walk(root);
  else { if (stat.size > MAX_TRANSFER) throw new Error("Transfer exceeds 512 MiB"); entries.push(path.basename(source)); }
  const archive = create({ cwd: root, portable: true, noMtime: true, follow: false, noDirRecurse: true }, entries);
  const chunks: Uint8Array[] = []; let size = 0;
  for await (const chunk of archive) {
    const data = Buffer.from(chunk); size += data.length;
    if (size > MAX_TRANSFER) throw new Error("Archive exceeds 512 MiB");
    chunks.push(data);
  }
  return { data: Buffer.concat(chunks), dockerfile };
}
export async function unpack(data: Uint8Array, destination: string): Promise<void> {
  if (data.byteLength > MAX_TRANSFER) throw new Error("Download exceeds 512 MiB");
  const root = path.resolve(destination);
  await mkdir(root, { recursive: true });
  let expanded = 0; let count = 0;
  const writer = extract({ cwd: root, strict: true, preservePaths: false,
    filter: (name, entry) => {
      if (++count > 100000) throw new Error("Too many archive entries");
      const target = path.resolve(root, name);
      if (path.isAbsolute(name) || (target !== root && !target.startsWith(root + path.sep))) throw new Error("Archive escapes destination");
      if (!("type" in entry) || (entry.type !== "File" && entry.type !== "Directory")) throw new Error("Archive contains a link or special file");
      expanded += entry.size ?? 0;
      if (expanded > MAX_TRANSFER) throw new Error("Expanded archive exceeds 512 MiB");
      return true;
    },
  });
  await new Promise<void>((resolve, reject) => {
    writer.on("error", reject); writer.on("close", resolve); writer.end(Buffer.from(data));
  });
}

File: src/opensandbox/sdk/files.py
"""Bounded, portable transfer archives and Docker build contexts."""

from __future__ import annotations

import io
import tarfile
from pathlib import Path

from pathspec.patterns.gitwildmatch import GitWildMatchPattern

from opensandbox.errors import ValidationError

MAX_TRANSFER = 512 * 1024 * 1024


class DockerIgnore:
    """Docker rules are rooted at the context and negations may restore children.

    Git's matching primitive handles escapes/globstars/character classes, but its
    default unanchored single-name rules and parent-directory pruning don't match
    Docker semantics. Anchor every rule and apply them in order to every path.
    """

    def __init__(self, content: str):
        self.rules = []
        for raw in content.splitlines():
            value = raw.strip()
            if not value or value.startswith("#"):
                continue
            excluded = not value.startswith("!")
            pattern = value if excluded else value[1:]
            pattern = pattern.removeprefix("./").strip("/")
            if pattern in {"", "."}:
                continue
            self.rules.append((GitWildMatchPattern("/" + pattern), excluded))

    def ignores(self, path: str):
        excluded = False
        for rule, value in self.rules:
            if rule.match_file(path) is not None:
                excluded = value
        return excluded


def pack(path: Path, *, context=False, limit=MAX_TRANSFER) -> tuple[bytes, str | None]:
    path = path.expanduser()
    if path.is_symlink():
        raise ValidationError("Transfers do not follow symlinks")
    path = path.resolve()
    if not path.exists():
        raise ValidationError(f"Local path does not exist: {path}")
    dockerfile = path.name if context and path.is_file() else "Dockerfile" if context else None
    root = path.parent if path.is_file() else path
    if context and (dockerfile is None or not (root / dockerfile).is_file()):
        raise ValidationError("Project context must contain a Dockerfile")
    ignore = DockerIgnore("")
    if context:
        ignore_path = root / (str(dockerfile) + ".dockerignore")
        if not ignore_path.exists():
            ignore_path = root / ".dockerignore"
        if ignore_path.exists():
            ignore = DockerIgnore(ignore_path.read_text())
    buffer = io.BytesIO()
    total = 0
    entries = list(root.rglob("*")) if context or path.is_dir() else [path]
    with tarfile.open(fileobj=buffer, mode="w") as archive:
        for item in sorted(entries):
            relative = item.relative_to(root).as_posix()
            # Credentials and repository/tool caches aren't part of a useful build context.
            if context and any(
                part in {".git", ".venv", ".cache", "node_modules"}
                for part in item.relative_to(root).parts
            ):
                continue
            if context and relative != dockerfile and ignore.ignores(relative):
                continue
            if item.is_symlink():
                raise ValidationError(f"Transfers do not follow symlinks: {relative}")
            if not item.is_dir() and not item.is_file():
                raise ValidationError(f"Cannot transfer special file: {relative}")
            total += item.stat().st_size if item.is_file() else 0
            if total > limit:
                raise ValidationError("Transfer exceeds 512 MiB")
            name = (
                relative
                if context
                else str(Path(path.name) / relative)
                if path.is_dir()
                else path.name
            )
            archive.add(item, arcname=name, recursive=False, filter=_clean_header)
    if len(buffer.getbuffer()) > limit:
        raise ValidationError("Transfer archive exceeds 512 MiB")
    return buffer.getvalue(), dockerfile


def _clean_header(info):
    info.uid = info.gid = 0
    info.uname = info.gname = ""
    info.mode &= 0o777
    return info


def unpack(data: bytes, destination: Path, *, limit=MAX_TRANSFER):
    if len(data) > limit:
        raise ValidationError("Download exceeds transfer limit")
    destination.mkdir(parents=True, exist_ok=True)
    total = 0
    with tarfile.open(fileobj=io.BytesIO(data)) as archive:
        members = archive.getmembers()
        if len(members) > 100000:
            raise ValidationError("Too many archive entries")
        for member in members:
            if not member.isfile() and not member.isdir():
                raise ValidationError("Archive contains a link or special file")
            total += member.size
            if total > limit:
                raise ValidationError("Expanded archive exceeds transfer limit")
            # Explicitly check existing symlinks as well as lexical traversal.
            target = (destination / member.name).resolve()
            if not target.is_relative_to(destination.resolve()) or Path(member.name).is_absolute():
                raise ValidationError("Archive escapes download destination")
        archive.extractall(destination, members=members, filter="data")
